### PR TITLE
Implement internal and external mode.

### PR DESCRIPTION
**Pull request by chris-wood**
Implements #154. Starting as a draft while we work through interop.

### CODE REVIEW

File: poc/opaque_core.sage [hunk 1]
```diff
@@ -63,30 +63,64 @@ class OPAQUECore(object):
...
+        Nh = self.config.hash().digest_size
+        auth_key = self.config.kdf.expand(random_pwd, envelope_nonce + _as_bytes("AuthKey"), Nh)
+        export_key = self.config.kdf.expand(random_pwd, envelope_nonce + _as_bytes("ExportKey"), Nh)
+        masking_key = self.derive_masking_key(random_pwd)
 
-        pt = secret_creds.serialize()
-        auth_data = cleartext_creds.serialize()
+        inner_env, client_public_key = self.build_inner_envelope(random_pwd, envelope_nonce, creds.skU)
+        cleartext_creds = self.create_cleartext_credentials(server_public_key, client_public_key, creds.idS, creds.idU)
+        auth_tag = self.config.mac.mac(auth_key, inner_env.serialize() + cleartext_creds.serialize())
```
Comment (kevinlewi) [inline poc/opaque_core.sage:110]: auth_tag needs to be computed over the mode and the nonce as well, not just the inner_env and cleartext_creds

File: poc/test_opaque_ake.sage [hunk 2]
```diff
@@ -180,103 +145,102 @@ def test_3DH():
...
+                # Protocol inputs
+                if idU:
+                    inputs["client_identity"] = to_hex(idU)
+                if idS:
+                    inputs["server_identity"] = to_hex(idS)
+                inputs["oprf_seed"] = to_hex(oprf_seed)
+                inputs["credential_identifier"] = to_hex(credential_identifier)
+                inputs["password"] = to_hex(pwdU)            
+                inputs["client_private_key"] = to_hex(skU_bytes)
+                inputs["client_public_key"] = to_hex(pkU_bytes)
```
Comment (kevinlewi) [inline poc/test_opaque_ake.sage:197]: One problem: when these are derived from the randomized_pwd seed, these are no longer inputs -- they are intermediate values. In particular, I don't think these were updated as their corresponding values when derived from the randomized_pwd seed, and so they end up not matching...

File: poc/opaque_core.sage [hunk 3]
```diff
@@ -63,30 +63,64 @@ class OPAQUECore(object):
...
+
+    def build_inner_envelope(self, random_pwd, envelope_nonce, client_private_key):
+        if self.config.mode == envelope_mode_internal:
+            Nsk = self.config.Nsk
+            seed = self.config.kdf.expand(random_pwd, envelope_nonce + _as_bytes("PrivateKey"), Nsk)
+            (_, client_public_key) = self.derive_group_key_pair(seed)
+            pk_bytes = self.config.group.serialize(client_public_key)
+            return (InnerEnvelope(), self.config.group.serialize(client_public_key))
+        if self.config.mode == envelope_mode_external:
+            secret_creds = SecretCredentials(client_private_key)
```
Comment (bytemare) [inline poc/opaque_core.sage:86]: We're dropping `SecretCredentials` and encrypt `client_private_key` directly

File: poc/opaque_core.sage (hunk 1 above)
Comment (bytemare) [reply poc/opaque_core.sage:110]: Agreed, that's where the vectors don't match :)

**Review by bytemare — CHANGES_REQUESTED** [review]

File: poc/test_opaque_ake.sage (hunk 2 above)
Comment (bytemare) [reply poc/test_opaque_ake.sage:197]: I confirm

**Comment by bytemare** [comment]
Something quite peculiar: P384 and P512 in the internal mode only have a public key in the registration upload vectors that doesn't correspond to either the client public key vector (that is false anyway) and the recovered public key from the password.

In all other vectors, the public key in the registration upload is correct.

**Pull request merged**